src/test/java/pages/BasePage.java

```diff
@@ -4,9 +4,11 @@
 
 abstract class BasePage {
     WebDriver driver;
+    String url;
 
-    BasePage(WebDriver driver) {
+    BasePage(WebDriver driver, String url) {
```

Comment:
Why you decided to pass String url in page object initialization? I don't think it's a good way to make such initialization.

You can create common method in BasePage:
    
public void openPage(String url) {
        driver.get(url);
    }

and just pass necessary constants on each page, for example:

ProductsPage:

    public void openPage() {
        openPage(SAUCE_DEMO_URL + PRODUCTS_URL);
    }

Similar we did on our latest lesson (29.12)

Reply:
done